### ISSUE TITLE
Database called recursively during migration

### ISSUE THREAD

**Issue by bartoszwilk** [agrosner/DBFlow#1520]
DBFlow Version: 4.2.4

Bug or Feature Request: Possible bug/question

Description:
Hi
What I'm trying to achieve is to recreate my table with saving all its data.
Here is my migration code:
```ruby    
@Migration(version = 2, database = VideostarDatabase::class)
    class ProvidersMigration : BaseMigration() {
        override fun migrate(database: DatabaseWrapper) {
            val models = SQLite.select().from(EpgTvProviderDbModel_EpgChannelDbModel::class.java).queryList(database)
            dropAndRecreateTable(database, EpgTvProviderDbModel_EpgChannelDbModel::class.java)
            models.save(database)
        }
    }
```
Regarding to [your advice about preventing recursive access](https://agrosner.gitbooks.io/dbflow/content/Migrations.html#prevent-recursive-access-to-the-db)  to db, I have used mentioned databaseWrapper to both select and save operation.

Unfortunately still getting the same 'getDatabase called recursively' error after reaching the first line of the migration block (the line with select operation). 
stacktrace:
```ruby
 java.lang.IllegalStateException: getDatabase called recursively
                                                                        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:234)
                                                                        at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:194)
                                                                        at com.raizlabs.android.dbflow.structure.database.FlowSQLiteOpenHelper.getDatabase(FlowSQLiteOpenHelper.java:62)
                                                                        at com.raizlabs.android.dbflow.config.DatabaseDefinition.getWritableDatabase(DatabaseDefinition.java:257)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ModelLoader.load(ModelLoader.java:36)
                                                                        at com.raizlabs.android.dbflow.sql.language.BaseModelQueriable.querySingle(BaseModelQueriable.java:69)
                                                                        at com.raizlabs.android.dbflow.sql.language.Where.querySingle(Where.java:249)
                                                                        at pl.wp.videostar.data.rdp.repository.impl.dbflow.epg_tv_provider.model.EpgTvProviderDbModel_EpgChannelDbModel_Table.loadFromCursor(EpgTvProviderDbModel_EpgChannelDbModel_Table.java:172)
                                                                        at pl.wp.videostar.data.rdp.repository.impl.dbflow.epg_tv_provider.model.EpgTvProviderDbModel_EpgChannelDbModel_Table.loadFromCursor(EpgTvProviderDbModel_EpgChannelDbModel_Table.java:23)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ListModelLoader.convertToData(ListModelLoader.java:77)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ListModelLoader.convertToData(ListModelLoader.java:15)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ModelLoader.load(ModelLoader.java:72)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ListModelLoader.load(ListModelLoader.java:61)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ListModelLoader.load(ListModelLoader.java:15)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ModelLoader.load(ModelLoader.java:60)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ListModelLoader.load(ListModelLoader.java:44)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ListModelLoader.load(ListModelLoader.java:15)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ModelLoader.load(ModelLoader.java:53)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ListModelLoader.load(ListModelLoader.java:37)
                                                                        at com.raizlabs.android.dbflow.sql.language.BaseModelQueriable.queryList(BaseModelQueriable.java:84)
                                                                        at pl.wp.videostar.data.rdp.repository.base.dbflow.VideostarDatabase$EpgTvProvidersChannelsManyToMany.migrate(VideostarDatabase.kt:134)
                                                                        at com.raizlabs.android.dbflow.structure.database.BaseDatabaseHelper.executeMigrations(BaseDatabaseHelper.java:171)
                                                                        at com.raizlabs.android.dbflow.structure.database.BaseDatabaseHelper.onUpgrade(BaseDatabaseHelper.java:56)
                                                                        at com.raizlabs.android.dbflow.structure.database.DatabaseHelperDelegate.onUpgrade(DatabaseHelperDelegate.java:79)
                                                                        at com.raizlabs.android.dbflow.structure.database.FlowSQLiteOpenHelper.onUpgrade(FlowSQLiteOpenHelper.java:84)
                                                                        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:299)
                                                                        at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:194)
                                                                        at com.raizlabs.android.dbflow.structure.database.FlowSQLiteOpenHelper.getDatabase(FlowSQLiteOpenHelper.java:62)
                                                                        at com.raizlabs.android.dbflow.config.DatabaseDefinition.getWritableDatabase(DatabaseDefinition.java:257)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ModelLoader.load(ModelLoader.java:36)
                                                                        at com.raizlabs.android.dbflow.sql.queriable.ListModelLoader.load(ListModelLoader.java:25)
                                                                        at com.raizlabs.android.dbflow.sql.language.BaseModelQueriable.queryList(BaseModelQueriable.java:61)
                                                                        at com.raizlabs.android.dbflow.sql.language.BaseTransformable.queryList(BaseTransformable.java:99)
                                                                        at com.raizlabs.android.dbflow.rx2.language.RXModelQueriableImpl$2.call(RXModelQueriableImpl.java:61)
                                                                        at com.raizlabs.android.dbflow.rx2.language.RXModelQueriableImpl$2.call(RXModelQueriableImpl.java:58)
                                                                        at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:35)
                                                                        at io.reactivex.Single.subscribe(Single.java:2700)
                                                                        at io.reactivex.internal.operators.single.SingleToObservable.subscribeActual(SingleToObservable.java:34)
                                                                            at io.reactivex.Observable.subscribe(Observable.java:10903)
                                                                        at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.subscribeInner(ObservableFlatMap.java:162)
                                                                        at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.onNext(ObservableFlatMap.java:139)
                                                                        at io.reactivex.internal.operators.single.SingleToObservable$SingleToObservableObserver.onSuccess(SingleToObservable.java:59)
                                                                        at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:81)
                                                                        at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
                                                                        at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:26)
                                                                        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
                                                                        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
                                                                        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
                                                                        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
                                                                        at java.lang.Thread.run(Thread.java:764)
```

Am I doing this migration properly?

**Comment by mseroczynski** [comment]
I'm having the same issue, can we get any support?

**Comment by artemtkachenko** [comment]
@bartoszwilk @mseroczynski 
Hello there

I had an identical issue a while ago. 

Unfortunately `DBFlow` was not designed properly for fetching the DAO while migration. 
It seems that a model what you are trying to load from wrapped-database contains a **@ForeignKey()** properties. That is why `querySingle()` is invoked all the time, this operation needs a `writable-database`. That is why 'getDatabase called recursively' error appears.

Check generated `_Table.java` file, method:
```
          @Override
          public final void loadFromCursor(FlowCursor cursor, %Table::class% model) {...}
```
You should avoid the default `load dao` implementation. I suggest using FlowCursor while migration.
E.g. 
```
        override fun migrate(database: DatabaseWrapper) {
           val cursor = SQLite.select().from(EpgTvProviderDbModel_EpgChannelDbModel::class.java).query(database)
           dropAndRecreateTable(database, EpgTvProviderDbModel_EpgChannelDbModel::class.java)
           val models = mutableListOf<%EntityModel%>
           if (cursor != null && cursor.count > 0) {
                cursor.move(-1)
                while (flowCursor.moveToNext()) {
                    //wrap cursor data onto your model with 
                   models.add(%EntityModel%.fromCursor(flowCursor))
                }
                cursor.close()
            }
            models.save(database)
        }
```
That's it.

**Comment by mseroczynski** [comment]
Wow, that's an awesome explanation and example. I've had to move forward with my tasks so I've just used *.sql file migration which worked correctly so far. Anyway, big thanks!

**Comment by agrosner** [comment]
i believe this is fixed in 5.0.0+ by always requiring a `DatabaseWrapper` when using db operations. This extends to generated code such as foreign keys.
